litestar/utils/sync.py

```
@@ -51,25 +54,30 @@ def __call__(self, *args: P.args, **kwargs: P.kwargs) -> Awaitable[T]:  # type:
         return sync_to_thread(self.func, *args, **kwargs)  # type: ignore[arg-type]
 
 
-class AsyncIteratorWrapper(Generic[T]):
+class AsyncGeneratorWrapper(AsyncGenerator[T, S]):
     """Asynchronous generator, wrapping an iterable or iterator."""
 
-    __slots__ = ("generator", "iterator")
+    __slots__ = ("_is_in_thread", "_original_iterator", "generator", "iterator")
 
     def __init__(self, iterator: Iterator[T] | Iterable[T]) -> None:
         """Take a sync iterator or iterable and yields values from it asynchronously.
 
         Args:
             iterator: A sync iterator or iterable.
         """
-        self.iterator = iterator if isinstance(iterator, Iterator) else iter(iterator)
+        self._is_in_thread = False
+        self._original_iterator = iterator
+        self.iterator = iter(iterator)
         self.generator = self._async_generator()
 
     def _call_next(self) -> T:
+        self._is_in_thread = True
         try:
             return next(self.iterator)
         except StopIteration as e:
             raise ValueError from e
+        finally:
+            self._is_in_thread = False
```

I should also mention that this synchronisation flag is not strictly required. Without it, the generator still manages to clean up but an exception is thrown. If it were up to me, I would just hide the exception with a `try-except`.